### ISSUE TITLE
OpenTelemetry.Instrumentation.Http not working in net6.0 with code loaded through reflection from different directory

### ISSUE THREAD

**Issue by m-mertz** [open-telemetry/opentelemetry-dotnet#4305]
# Bug Report

List of [all OpenTelemetry NuGet
packages](https://www.nuget.org/profiles/OpenTelemetry) and version that you are
using (e.g. `OpenTelemetry 1.0.2`):

* OpenTelemetry 1.4.0
* OpenTelemetry.Exporter.Console 1.4.0
* OpenTelemetry.Instrumentation.Http 1.0.0-rc9.14

Runtime version: net6.0

## Symptom

We have a setup consisting of the following pieces:
* A net6.0 console app as a "job runtime".
* A netstandard2.0 library containing OpenTelemetry SDK initialization code for reuse across "job implementations".
* "Job implementations" in separate net6.0 libraries.

The repro project is a massive simplification of our actual code which preserves this overall structure.

Flow:
* The console app loads one of the job implementations through reflection from a different directory and runs the job class: https://github.com/m-mertz/OTelHttpRepro/blob/main/OTelHttpRepro/Program.cs
* The job class configures the OpenTelemetry.Instrumentation.Http package and makes an HttpClient.GetAsync call which should get captured: https://github.com/m-mertz/OTelHttpRepro/blob/main/JobLib/Job.cs

**What is the expected behavior?**

* HTTP Get span gets printed by Console Exporter for the HttpClient.GetAsync call in the program.

**What is the actual behavior?**

* HTTP Get span does not get printed.
* Changing the TargetFramework from net6.0 to net7.0 "fixes" the issue.
* It also works with net6.0 when running the job directly through project reference instead of loading it through reflection from binaries in a different directory.

## Reproduce

* Clone this repo: https://github.com/m-mertz/OTelHttpRepro
* Build + run the OTelHttpRepro project.
* Main method also contains instructions for the two "fixes" mentioned above: https://github.com/m-mertz/OTelHttpRepro/blob/main/OTelHttpRepro/Program.cs

**Comment by vishweshbankwar** [comment]
@m-mertz I am not familiar with the way you are using reflection to load the assemblies. But I am guessing it has to do something with the way `System.Diagnostics.DiagnosticSource` is getting loaded. In case of `net6.0` it is loaded twice (possibly in two different load contexts? again I am not sure why)

![image](https://user-images.githubusercontent.com/17655898/228126800-f30a2651-29f1-44ed-b80a-d9153e0c8b32.png)

Unless this ^ is expected, would recommend to find the root cause for that first. 
In case of net7.0 (working case) it is loaded only once.

**Comment by m-mertz** [comment]
Thanks @vishweshbankwar ! Based on our debugging it looks like one of those is loaded from the JobLib bin folder (where we expect dependencies to be loaded from), and the other one is loaded from .NET 6 runtime installation path. So it may be the mismatch between those two versions (v7 from bin folder as OpenTelemetry dependency, and v6 from .NET 6 runtime) that is causing the issue.

_(This is after removing JobLib reference from OTelHttpRepro project, which is only there to demo the mitigation option "don't use reflection" in this repro setup.)_

I'll see if I can figure out why v6 of System.Diagnostics.DiagnosticSource is loaded in the first place.

**Comment by shawner18** [comment]
I also ran into the same issue using the AWS lambda testing tool. The test tool is loading the 6.0 System.Diagnostics.DiagnosticsSource dll, and the later versions of opentelemetry reference the 7.0 assemblies. I saw both assemblies listed in the debugger like @vishweshbankwar presented.

I was original investigating why a lambda we wrote was not including the headers but logging the trace. I can assume AWS lambdas are loading the executing code in a similar way as the test tool.
I was not seeing any other side effects of this except that the injection of trace headers were missing on outgoing calls. 

This combination of references brought the header injection back. Though I am not sure what issues or features I might be missing having to go back so far on some of these.
![image](https://user-images.githubusercontent.com/32966556/234627767-b43a793d-afb4-4964-8d25-b36233d16ae8.png)
